Feat: Add `servicename` field to the Prometheus server and agent CRDs

mviswanathsai opened the pull request:
## Description

[#6060](https://github.com/prometheus-operator/prometheus-operator/issues/6060) relates to exposing services with custom service names since  "governing" services don’t play well when the Prometheus-es in the same namespace need to expose different sets of ports, which can only be done by using different services. 
This PR adds a field called `servicename` where the user can specify the name of the service that will expose that particular set of Prometheus instances which are governed by the Prometheus CR. 

The behavior is as follows: 
1. If .spec.serviceName is defined and the service doesn't exist, the operator fails the reconciliation.
2. If the service exists, it exposes the respective Promethei.
3. If the .spec.serviceName is not defined, a regular governing service is exposed.

It is up to the user to expose a service that suits their needs. However, the operator does help in preventing users from linking the Prometheus resource with a service that does not select it. 

Closes #6060 

## Type of change

_What type of changes does your code introduce to the Prometheus operator? Put an `x` in the box that apply._

- [ ] `CHANGE` (fix or feature that would cause existing functionality to not work as expected)
- [x] `FEATURE` (non-breaking change which adds functionality)
- [ ] `BUGFIX` (non-breaking change which fixes an issue)
- [ ] `ENHANCEMENT` (non-breaking change which improves existing functionality)
- [ ] `NONE` (if none of the other choices apply. Example, tooling, build system, CI, docs, etc.)

## Verification
<!-- How you tested it? How do you know it works? -->
Please check the [Prometheus-Operator testing guidelines](../TESTING.md) for recommendations about automated tests.

## Changelog entry

_Please put a one-line changelog entry below. This will be copied to the changelog file during the release process._

<!-- 
Your release note should be written in clear and straightforward sentences. Most often, users aren't familiar with
the technical details of your PR, so consider what they need to know when you write your release note.

Some brief examples of release notes:
- Add metadataConfig field to the Prometheus CRD for configuring how remote-write sends metadata information.
- Generate correct scraping configuration for Probes with empty or unset module parameter.
-->

```
Add `servicename` field to common Prometheus fields
```

simonpasquier commented on pkg/apis/monitoring/v1/prometheus_types.go:
```diff
@@ -701,6 +701,13 @@ type CommonPrometheusFields struct {
 	// +listMapKey=name
 	ScrapeClasses []ScrapeClass `json:"scrapeClasses,omitempty"`
+
+	// The service name by which we wish to expose the Prometheus pods
+	// governed by the given Prometheus CR.
```
```suggestion
	// The name of the service name used by the underlying StatefulSet as its governing service. If defined, it must be created before the Prometheus object.
	// If empty, the operator will create and manage a service named `prometheus-operated`.
	// See https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/#stable-network-id for more details.
```

simonpasquier commented on pkg/prometheus/agent/operator.go:
```diff
@@ -605,10 +605,35 @@ func (c *Operator) sync(ctx context.Context, key string) error {
+	// the reconciliation.
+	// If the ServiceName is not specified, create a governing service if it doesn't exist.
+	// Also, ensure that the Prometheus instance is selected by the service. If not, fail the reconciliation.
+	// TODO[mviswanathsai]: is it possible to refactor this (and other) code repetitions between the Prometheus server and agent?
```
yes please :)
side suggestion: please do in a preliminary PR.

mviswanathsai replied:
Should we complete this PR and then move onto the refactor? Seems like the refactor might be a bit of work.

simonpasquier replied:
I'd rather do it before this PR otherwise there's a risk we never get to it.

mviswanathsai replied:
Got it!

mviswanathsai replied:
Can you give me pointers on where to start with the refactor? @simonpasquier

simonpasquier replied:
Sorry I was confused: it should be done in this PR. The logic when `p.Spec.ServiceName != nil` should be similar between agent and server.

simonpasquier commented on pkg/prometheus/agent/operator.go:
```diff
@@ -605,10 +605,35 @@ func (c *Operator) sync(ctx context.Context, key string) error {
+			return fmt.Errorf("failed to get label selector while synchronizing service: %w", err)
+		}
+
+		if svcSelector.Matches(labels.Set(p.Labels)) {
```
you want to check against the pod's labels, not the Prometheus object label.

mviswanathsai replied:
Got it.

simonpasquier replied:
this is still not correct: the Kubernetes service's selector is used to select pods based on their labels. In practice you need to use `makeSelectorLabels()` (the function could use a more expressive name, e.g. `selectorForPrometheusPods`).

mviswanathsai replied:
Not sure to follow here, I do use `makeSelectorLabels` and then check against the label selectors present in the service.

mviswanathsai commented:
The e2e test I'd written fails, I just wanted to push it to get some opinions on it as well.
CC: @slashpai @ArthurSens

mviswanathsai commented:
PTAL @simonpasquier. I also have a few things to discuss regarding the e2e tests, and I have added some concerns as `TODO` comments in the code.

simonpasquier commented on pkg/prometheus/common.go:
```diff
@@ -594,3 +598,31 @@ func MakeProbes(
+	// Get the user defined label selectors from the service.
+	svcSelector, err := metav1.LabelSelectorAsSelector(&metav1.LabelSelector{MatchLabels: svc.Spec.Selector})
+	if err != nil {
+		return fmt.Errorf("failed to get label selector while synchronizing service: %w", err)
```
IIUC this should never happen in real life.
```suggestion
		return fmt.Errorf("failed to convert service selector: %w", err)
```

mviswanathsai replied:
I have made the change, but could you elaborate a bit on that statement?

simonpasquier replied:
I mean to say that the service shouldn't have an invalid label selector value (how would it be allowed by the API in the first place?). But it's better to handle the error in any case.

simonpasquier commented on pkg/prometheus/common.go:
```diff
@@ -594,3 +598,31 @@ func MakeProbes(
+	// Check if the svcSelector is the same as the Selector used to match the pods handled by this
+	// Prometheus resource.
+	if !svcSelector.Matches(labels.Set(selectorLabels)) {
+		return fmt.Errorf("service %s/%s does not select prometheus server %s", prometheusNS, *serviceName, prometheusName)
```
(suggestion) let's print out the details.
```suggestion
		return fmt.Errorf("service %s/%s  with selector %q does not select Prometheus pods %s with labels %q", prometheusNS, *serviceName, prometheusName, svcSelector.String(), labels.Set(selectorLabels).String())
```

mviswanathsai replied:
Changed it a bit
```
		return fmt.Errorf("service %s/%s with selector %s does not select Prometheus sset pods %s/%s with labels %s",
			prometheusNS, serviceName, svcSelector.String(), prometheusNS, prometheusName, labels.Set(selectorLabels).String())
```

simonpasquier commented on pkg/prometheus/common.go:
```diff
@@ -594,3 +598,31 @@ func MakeProbes(
+// If a ServiceName is specified, check that the service exists in the namespace. If it does not exist, fail
+// the reconciliation.
+// Also, ensure that the Prometheus instance is selected by the service. If not, fail the reconciliation.
+func CheckCustomService(ctx context.Context, serviceName *string, prometheusNS string, prometheusName string, svcClient clientv1.ServiceInterface, selectorLabels map[string]string) error {
```
What if we pass a PrometheusInterface parameter instead? It would simplify the function signature.

mviswanathsai replied:
I think at the time of writing, my POV was that do we need to pass the entire interface when we only need 3 string values from it. But no strong opinions on this one.

simonpasquier commented on test/e2e/prometheus_test.go:
```diff
@@ -5169,6 +5169,77 @@ func testPrometheusStatusScale(t *testing.T) {
 	}
 }
 
+// TODO: Also make sure no governing service was created?
```
I'd be ok not checking this.

simonpasquier commented on test/e2e/prometheus_test.go:
```diff
@@ -5169,6 +5169,77 @@ func testPrometheusStatusScale(t *testing.T) {
+		t.Fatal(err)
+	}
+
+	// TODO: Maybe we can just send a http request to the Prometheus pod to see if it is active?
```
IIUC the existing test is enough: WaitForHealthyTargets() will target the custom service so if it replies ok, you know that everything is wired properly.

mviswanathsai replied:
I thought perhaps we could do away with the dummy target creation, but since it does mimic real life scenarios I think it should be good.

simonpasquier commented on test/e2e/prometheus_test.go:
```diff
@@ -5169,6 +5169,77 @@ func testPrometheusStatusScale(t *testing.T) {
+// What do we want to do? we want to make sure that when you are creating your own service and
+// the service is actually selected. But this is pretty self explanatory.
+// Something that we also want to check for is to make sure that if the service does not select
+// our prometheus, and the prometheus selects our service then we reject such a config.
```
Yes this would useful to test the error cases.

mviswanathsai replied:
Forgot to remove this, mb!

mviswanathsai commented:
Random issues kept popping up with the local e2e tests (not concerning the actual test itself) which lead to failures over and over again. So I decided to just push the changes and then deal with it.

mviswanathsai commented:
I think this is good for a review now @simonpasquier

simonpasquier commented on pkg/apis/monitoring/v1/prometheus_types.go:
```diff
@@ -821,6 +821,14 @@ type CommonPrometheusFields struct {
 	// +optional
 	TSDB *TSDBSpec `json:"tsdb,omitempty"`
+
+	// The name of the service name used by the underlying StatefulSet as its governing service.
```
(nit) there are multiple statefulsets in case of sharding.
```suggestion
	// The name of the service name used by the underlying StatefulSet(s) as the governing service.
```

simonpasquier commented on pkg/apis/monitoring/v1/prometheus_types.go:
```diff
@@ -821,6 +821,14 @@ type CommonPrometheusFields struct {
+	// If defined, it must be created before the Prometheus object in the same namespace and it must define a selector that will match the pod labels.
+	// If empty, the operator will create and manage a service named `prometheus-operated`.
+	// When deploying multiple Prometheus/PrometheusAgent resources in the same namespace, it is recommended to specify a different value for each.
+	// See https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/#stable-network-id for more details.	// +optional
```
```suggestion
	// See https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/#stable-network-id for more details.
```

simonpasquier commented on pkg/apis/monitoring/v1/prometheus_types.go:
```diff
@@ -821,6 +821,14 @@ type CommonPrometheusFields struct {
+	// If empty, the operator will create and manage a service named `prometheus-operated`.
+	// When deploying multiple Prometheus/PrometheusAgent resources in the same namespace, it is recommended to specify a different value for each.
+	// See https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/#stable-network-id for more details.	// +optional
+	// +optional
```
We'll need a validation on the length too (>= 1).

simonpasquier commented on pkg/prometheus/agent/operator.go:
```diff
@@ -714,10 +714,18 @@ func (c *Operator) syncStatefulSet(ctx context.Context, key string, p *monitorin
+			return fmt.Errorf("synchronizing custom service failed: %w", err)
+		}
+	} else {
+		// If the ServiceName is not specified, create a governing service if one doesn't already exist.
```
```suggestion
		// If the ServiceName is not specified, the controller manages the governing service.
```

simonpasquier commented on pkg/prometheus/server/operator.go:
```diff
@@ -787,10 +787,22 @@ func (c *Operator) sync(ctx context.Context, key string) error {
 	}
 
-	// Create governing service if it doesn't exist.
+	if err := c.createOrUpdateThanosConfigSecret(ctx, p); err != nil {
+		return fmt.Errorf("failed to reconcile Thanos config secret: %w", err)
+	}
```
mistake?
```suggestion
```

mviswanathsai replied:
Might've accidentally smashed the key binding

simonpasquier commented on pkg/prometheus/common.go:
```diff
@@ -439,3 +443,32 @@ func BuildWebconfig(
+// If a ServiceName is specified, check that the service exists in the namespace. If it does not exist, fail
+// the reconciliation.
+// Also, ensure that the Prometheus instance is selected by the service. If not, fail the reconciliation.
+func CheckCustomService(ctx context.Context, serviceName string, prometheusNS string, prometheusName string, svcClient clientv1.ServiceInterface, selectorLabels map[string]string) error {
```
If the function receives a `PrometheusInterface` argument + svcClient + selectorLabels, it could do:
* Verify that the service exists and matches the pods if p.Spec.ServiceName isn't nil
* Otherwise create/update the prometheus-operated service.

mviswanathsai replied:
I agree, that way we could do everything related to the service creation within this one function.

simonpasquier commented on pkg/apis/monitoring/v1/prometheus_types.go:
```diff
@@ -862,6 +862,15 @@ type CommonPrometheusFields struct {
 
+	// The name of the service name used by the underlying StatefulSet(s) as the governing service.
+	// If defined, it must be created before the Prometheus object in the same namespace and it must define a selector that will match the pod labels.
+	// If empty, the operator will create and manage a service named `prometheus-operated`.
```
```suggestion
	// If empty, the operator will create and manage a headless service named `prometheus-operated` for Prometheus resources or `prometheus-agent-operated` for PrometheusAgent resources.
```

simonpasquier commented on pkg/apis/monitoring/v1/prometheus_types.go:
```diff
@@ -862,6 +862,15 @@ type CommonPrometheusFields struct {
 	TSDB *TSDBSpec `json:"tsdb,omitempty"`
 
+	// The name of the service name used by the underlying StatefulSet(s) as the governing service.
+	// If defined, it must be created before the Prometheus object in the same namespace and it must define a selector that will match the pod labels.
```
```suggestion
	// If defined, it must be created before the Prometheus/PrometheusAgent resource in the same namespace and it must define a selector that will match the pod labels.
```

mviswanathsai commented:
Should we take the de-duplication commit up here? or maybe have it in a separate PR with tests (if necessary). I don't have a strong opinion but I would be leaning a bit towards the separate PR.

mviswanathsai commented:
```
Error: test/e2e/prometheus_test.go:5338:89: cannot use "web" (untyped string constant) as *string value in struct literal
```
Not sure this error is relevant? the type for Port is `string` AFAIK.

simonpasquier commented:
that's a compilation error in the e2e tests.

mviswanathsai commented:
Yes, the e2e tests compiled just fine earlier and nothing has changed since AFAIK...

mviswanathsai commented:
This is also how the type has been used in other parts of the code
```go
					PodMetricsEndpoints: []monitoringv1.PodMetricsEndpoint{
						{
							Port:     "web",
							Interval: "30s",
// Other specifications
                                                 }
```

simonpasquier commented:
Hmm it's strange. Maybe rebase on top of main?

simonpasquier commented:
you'll need to rebase since #7162 merged

mviswanathsai commented:
I rebased and changed the logic a bit. The `CheckGoverningService` function was initially doing two things: validation, or creation of a new service. Now, I decided to keep it in the same function because it ultimately has to do with the "governing" function for the statefulset, but come to think of it... I cannot really name the function properly.

It sounds silly but I decided if I cannot name the function then it is probably doing a couple of things that shouldn't be done by one function. Of course, it can, but it didn't really sit right for me at least in terms of readability. Let me know what you think about this @simonpasquier.

simonpasquier commented on pkg/apis/monitoring/v1/prometheus_types.go:
```diff
@@ -862,6 +862,16 @@ type CommonPrometheusFields struct {
 	TSDB *TSDBSpec `json:"tsdb,omitempty"`
 
+	// The name of the service name used by the underlying StatefulSet(s) as the governing service.
+	// If defined, it must be created before the Prometheus/PrometheusAgent resource in the same namespace and it must define a selector that will match the pod labels.
```
(nit) just to be explicit
```suggestion
	// If defined, the Service must be created before the Prometheus/PrometheusAgent resource in the same namespace and define a selector that matches the pod labels.
```

simonpasquier commented on pkg/prometheus/common.go:
```diff
@@ -474,3 +478,39 @@ func BuildStatefulSetService(name string, selector map[string]string, p monitori
 	return svc
 }
+
+// This function is responsible for the following:
+// If a ServiceName is specified, verify that the custom governing service exists in the namespace.
+// If it does not exist, fail the reconciliation.
+// Also, ensure that the Prometheus instance is selected by the custom governing service.
+// If it is not selected, fail the reconciliation.
+// If a ServiceName is not specified, create or update the default governing service.
```
The comment needs an update.

simonpasquier commented on pkg/prometheus/common.go:
```diff
@@ -474,3 +478,39 @@ func BuildStatefulSetService(name string, selector map[string]string, p monitori
+// Also, ensure that the Prometheus instance is selected by the custom governing service.
+// If it is not selected, fail the reconciliation.
+// If a ServiceName is not specified, create or update the default governing service.
+func CheckCustomGoverningService(ctx context.Context, prometheus monitoringv1.PrometheusInterface, svcClient clientv1.ServiceInterface, selectorLabels map[string]string) error {
```
(nit)
```suggestion
func EnsureGoverningService(ctx context.Context, prometheus monitoringv1.PrometheusInterface, svcClient clientv1.ServiceInterface, selectorLabels map[string]string) error {
```

simonpasquier commented on pkg/prometheus/common.go:
```diff
@@ -474,3 +478,39 @@ func BuildStatefulSetService(name string, selector map[string]string, p monitori
+	prometheusNS := prometheus.GetObjectMeta().GetNamespace()
+	prometheusName := prometheus.GetObjectMeta().GetName()
+
+	if serviceName == nil {
+		return nil
+	}
```
This would be a programmatic bug.

simonpasquier commented on pkg/prometheus/common.go:
```diff
@@ -474,3 +478,39 @@ func BuildStatefulSetService(name string, selector map[string]string, p monitori
+	// Check if the custom governing service is defined in the same namespace and selects the Prometheus pod.
+	svc, err := svcClient.Get(ctx, *serviceName, metav1.GetOptions{})
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return fmt.Errorf("custom governing service %s/%s does not exist", prometheusNS, *serviceName)
+		}
+
```
I wouldn't have a special handling for not-found errors.
```suggestion
```

simonpasquier commented on pkg/prometheus/common.go:
```diff
@@ -474,3 +478,39 @@ func BuildStatefulSetService(name string, selector map[string]string, p monitori
+	}
+
+	if !svcSelector.Matches(labels.Set(selectorLabels)) {
+		return fmt.Errorf("custom governing service %s/%s with selector %s does not select Prometheus sset pods %s/%s with labels %s",
```
```suggestion
		return fmt.Errorf("custom governing service %s/%s with selector %s does not select Prometheus pods %s/%s with labels %s",
```

simonpasquier commented on pkg/prometheus/common.go:
```diff
@@ -474,3 +478,39 @@ func BuildStatefulSetService(name string, selector map[string]string, p monitori
+// Also, ensure that the Prometheus instance is selected by the custom governing service.
+// If it is not selected, fail the reconciliation.
+// If a ServiceName is not specified, create or update the default governing service.
+func CheckCustomGoverningService(ctx context.Context, prometheus monitoringv1.PrometheusInterface, svcClient clientv1.ServiceInterface, selectorLabels map[string]string) error {
```
it should be possible to write unit tests for this function.

mviswanathsai replied:
Could you point me to an example that uses mock clients?

simonpasquier replied:
sure!

https://github.com/prometheus-operator/prometheus-operator/blob/b2a6a6ed8c33ee9a9a8eedd2809fbbe5895e2aba/pkg/prometheus/resource_selector_test.go#L1079-L1102

mviswanathsai replied:
Thanks!

simonpasquier commented on test/e2e/prometheus_test.go:
```diff
@@ -5300,6 +5300,70 @@ func testPrometheusStatusScale(t *testing.T) {
+		t.Fatal(fmt.Errorf("failed to create app service: %w", err))
+	}
+
+	pm := framework.MakeBasicPodMonitor(name)
```
do we really need to create a pod monitor and check the targets?

mviswanathsai replied:
I think I rushed it at the time, maybe we could use `framework.PrometheusSVCGetRequest()`.

simonpasquier commented on test/e2e/prometheus_test.go:
```diff
@@ -5300,6 +5300,70 @@ func testPrometheusStatusScale(t *testing.T) {
+	// Ensure that governing service was not created.
+	governingServiceName := "prometheus-operated"
+	if _, err := framework.KubeClient.CoreV1().Services(ns).Get(context.Background(), governingServiceName, metav1.GetOptions{}); err != nil {
+		if !apierrors.IsNotFound(err) {
```
we need to ensure that err != nil and apierrors.IsNotFound(err) is true.

simonpasquier commented on pkg/prometheus/common.go:
```diff
@@ -474,3 +477,36 @@ func BuildStatefulSetService(name string, selector map[string]string, p monitori
+// If it is not selected, fail the reconciliation.
+func EnsureCustomGoverningService(ctx context.Context, prometheus monitoringv1.PrometheusInterface, svcClient clientv1.ServiceInterface, selectorLabels map[string]string) error {
+	serviceName := prometheus.GetCommonPrometheusFields().ServiceName
+	prometheusNS := prometheus.GetObjectMeta().GetNamespace()
```
(nit)
```suggestion
	namespace := prometheus.GetObjectMeta().GetNamespace()
```

simonpasquier commented on pkg/prometheus/common.go:
```diff
@@ -474,3 +477,36 @@ func BuildStatefulSetService(name string, selector map[string]string, p monitori
+// Prometheus/PrometheusAgent statefulsets.
+// If it is not selected, fail the reconciliation.
+func EnsureCustomGoverningService(ctx context.Context, prometheus monitoringv1.PrometheusInterface, svcClient clientv1.ServiceInterface, selectorLabels map[string]string) error {
+	serviceName := prometheus.GetCommonPrometheusFields().ServiceName
```
(nit)
```suggestion
	name := prometheus.GetCommonPrometheusFields().ServiceName
```

simonpasquier commented on pkg/prometheus/common.go:
```diff
@@ -474,3 +477,36 @@ func BuildStatefulSetService(name string, selector map[string]string, p monitori
+	}
+
+	if !svcSelector.Matches(labels.Set(selectorLabels)) {
+		return fmt.Errorf("custom governing service %s/%s with selector %s does not select Prometheus pods %s/%s with labels %s",
```
(nit) I don't think that we need to repeat the prometheus resource's name. It should already be present in the logged information

mviswanathsai replied:
Got it.

simonpasquier commented on pkg/prometheus/common.go:
```diff
@@ -474,3 +477,36 @@ func BuildStatefulSetService(name string, selector map[string]string, p monitori
+
+// This function is responsible for the following:
+//
+// If `ServiceName` is specified, verify that a service with the same name exists and is in the same namespace
```
an invariant of the function is that ServiceName isn't nil.
```suggestion
// Verify that the service exists in the Prometheus resource's namespace.
```

simonpasquier commented on pkg/prometheus/common.go:
```diff
@@ -474,3 +477,36 @@ func BuildStatefulSetService(name string, selector map[string]string, p monitori
+// If the ServiceName is specified and a service with the same name exists in the same namespace as the
+// Prometheus/PrometheusAgent resource, ensure that the custom governing service's selector matches the
+// Prometheus/PrometheusAgent statefulsets.
+// If it is not selected, fail the reconciliation.
```
```suggestion
// If it is not selected, fail the reconciliation.
// Warning: the function will panic if the resource's ServiceName is nil.
```

simonpasquier commented on pkg/prometheus/server/operator.go:
```diff
@@ -793,28 +793,33 @@ func (c *Operator) sync(ctx context.Context, key string) error {
+			return fmt.Errorf("validating custom service failed: %w", err)
+		}
+	} else {
+		// If the ServiceName is not specified, create a governing service if one doesn't already exist.
```
```suggestion
		// Reconcile the default governing service.
```

simonpasquier commented on pkg/prometheus/server/operator.go:
```diff
@@ -793,28 +793,33 @@ func (c *Operator) sync(ctx context.Context, key string) error {
-	)
+	if p.Spec.ServiceName != nil {
+		if err := prompkg.EnsureCustomGoverningService(ctx, p, svcClient, selectorLabels); err != nil {
+			return fmt.Errorf("validating custom service failed: %w", err)
```
(nit) We can probably avoid wrapping the returned error which has enough information already IMHO.
```suggestion
			return err
```

simonpasquier commented on pkg/apis/monitoring/v1/prometheus_types.go:
```diff
@@ -862,6 +862,16 @@ type CommonPrometheusFields struct {
 	TSDB *TSDBSpec `json:"tsdb,omitempty"`
 
+	// The name of the service name used by the underlying StatefulSet(s) as the governing service.
+	// If defined, the Service  must be created before the Prometheus/PrometheusAgent resource in the same namespace and it must define a selector that will matches the pod labels.
```
```suggestion
	// If defined, the Service  must be created before the Prometheus/PrometheusAgent resource in the same namespace and it must define a selector that  matches the pod labels.
```

simonpasquier commented on pkg/prometheus/common.go:
```diff
@@ -474,3 +477,32 @@ func BuildStatefulSetService(name string, selector map[string]string, p monitori
+
+	svcSelector, err := metav1.LabelSelectorAsSelector(&metav1.LabelSelector{MatchLabels: svc.Spec.Selector})
+	if err != nil {
+		return fmt.Errorf("failed to get the selector labels for custom governing service %s/%s: %w", namespace, serviceName, err)
```
(nit)
```suggestion
		return fmt.Errorf("failed to parse the selector labels for custom governing service %s/%s: %w", namespace, serviceName, err)
```

simonpasquier commented on pkg/prometheus/common.go:
```diff
@@ -474,3 +477,32 @@ func BuildStatefulSetService(name string, selector map[string]string, p monitori
+	}
+
+	if !svcSelector.Matches(labels.Set(selectorLabels)) {
+		return fmt.Errorf("custom governing service %s/%s with selector %s does not select Prometheus/PrometheusAgent pods with labels %s",
```
```suggestion
		return fmt.Errorf("custom governing service %s/%s with selector %q does not select Prometheus/PrometheusAgent pods with labels %q",
```

simonpasquier commented on pkg/prometheus/agent/operator.go:
```diff
@@ -714,15 +714,24 @@ func (c *Operator) syncStatefulSet(ctx context.Context, key string, p *monitorin
-	if _, err := k8sutil.CreateOrUpdateService(ctx, c.kclient.CoreV1().Services(p.Namespace), svc); err != nil {
-		return fmt.Errorf("synchronizing governing service failed: %w", err)
+	svcClient := c.kclient.CoreV1().Services(p.Namespace)
+	selectorLabels := makeSelectorLabels(p.Name)
```
(nit) this could be moved in the first if block

simonpasquier commented on pkg/prometheus/server/operator.go:
```diff
@@ -793,28 +793,33 @@ func (c *Operator) sync(ctx context.Context, key string) error {
-		return fmt.Errorf("failed to reconcile Thanos config secret: %w", err)
-	}
+	svcClient := c.kclient.CoreV1().Services(p.Namespace)
+	selectorLabels := makeSelectorLabels(p.Name)
```
same here

mviswanathsai commented:
Addressed comments, not sure why the other e2e test is failing though?

simonpasquier commented:
PrometheusAgentSSetServiceName fails with an error which might be transient:
````
=== CONT  TestAllNS/y/PrometheusAgentSSetServiceName
    prometheusagent_test.go:623: 
        	Error Trace:	/home/runner/work/prometheus-operator/prometheus-operator/test/e2e/prometheusagent_test.go:623
        	Error:      	Received unexpected error:
        	            	the server is currently unable to handle the request (get services http:test-agent-servicename-service:web)
        	Test:       	TestAllNS/y/PrometheusAgentSSetServiceName
````

TestAllNS/y/PromWebWithThanosSidecar looks more genuine but I can't tell what's happening except that the init config reloader fails to start:

````
=== NAME  TestAllNS/y/PromWebWithThanosSidecar
    prometheus_test.go:3801: Creating prometheus failed: waiting for 1 Prometheus instances timed out (basic-prometheus): prometheus allns-y-promwebwiththanossidecar-so4dvz-1-5a129d16/basic-prometheus failed to become available: client rate limiter Wait returned an error: context deadline exceeded: context deadline exceeded
allns-y-promwebwiththanossidecar-so4dvz-1-5a129d16: failed to get pod logs: failed to retrieve logs of container "prometheus" (pod allns-y-promwebwiththanossidecar-so4dvz-1-5a129d16/prometheus-basic-prometheus-0): the server rejected our request for an unknown reason (get pods prometheus-basic-prometheus-0)
````

mviswanathsai commented:
> PrometheusAgentSSetServiceName fails with an error which might be transient:

Ah, this is probably because we are sending the request before the Prometheus(Agent) becomes available. I made the change with the Prometheus test, forgot the agent test.

mviswanathsai commented:
> TestAllNS/y/PromWebWithThanosSidecar looks more genuine but I can't tell what's happening except that the init config reloader fails to start:

This was happening because I accidentally removed the code that creates the Thanossidecar's webconfig secret. Yikes!

mviswanathsai commented:
Not entirely sure why the test is failing. But can you review the PR anyway @simonpasquier ?

simonpasquier commented on test/e2e/prometheus_test.go:
```diff
@@ -5300,6 +5300,56 @@ func testPrometheusStatusScale(t *testing.T) {
+		},
+	}
+
+	_, err := framework.CreateOrUpdateService(context.Background(), ns, svc)
```
we don't really need a create-or-update function?
```suggestion
	_, err := framework..KubeClient.CoreV1().Services(ns).Create(ctx, svc, metav1.CreateOptions{})
```

simonpasquier commented on test/e2e/prometheus_test.go:
```diff
@@ -5300,6 +5300,56 @@ func testPrometheusStatusScale(t *testing.T) {
+	framework.SetupPrometheusRBAC(context.Background(), t, testCtx, ns)
+
+	p := framework.MakeBasicPrometheus(ns, name, name, 1)
+	p.Spec.ServiceName = ptr.To(fmt.Sprintf("%s-service", name))
```
```suggestion
	p.Spec.ServiceName = ptr.To(svc.Name)
```

simonpasquier commented on test/e2e/prometheus_test.go:
```diff
@@ -5300,6 +5300,56 @@ func testPrometheusStatusScale(t *testing.T) {
+	_, err = framework.CreatePrometheusAndWaitUntilReady(context.Background(), ns, p)
+	require.NoError(t, err)
+
+	success, err := framework.BasicQueryWorking(context.Background(), ns, svc.Name)
```
why not use `framework.GetActiveTargets` here too?

simonpasquier reviewed (commented):
just a few comments on the e2e tests otherwise lgtm, thanks!

simonpasquier commented on test/framework/service.go:
```diff
@@ -71,6 +71,35 @@ func (f *Framework) CreateOrUpdateServiceAndWaitUntilReady(ctx context.Context,
 	return finalizerFn, nil
 }
 
+func (f *Framework) CreateOrUpdateService(ctx context.Context, namespace string, service *v1.Service) (FinalizerFn, error) {
```
I guess that we don't need this function anymore?

mviswanathsai replied:
Yep! missed it somehow. Also went ahead and reverted back the change to `basicQueryWorking`.

simonpasquier commented on test/framework/prometheus.go:
We don't need this change anymore?

simonpasquier commented on test/e2e/prometheusagent_test.go:
```diff
@@ -579,6 +580,57 @@ func testPrometheusAgentDaemonSetSelectPodMonitor(t *testing.T) {
+	framework.SetupPrometheusRBAC(context.Background(), t, testCtx, ns)
+
+	p := framework.MakeBasicPrometheusAgent(ns, name, name, 1)
+	p.Spec.ServiceName = ptr.To(fmt.Sprintf("%s-service", name))
```
(nit)
```suggestion
	p.Spec.ServiceName = &svc.Name
```

simonpasquier commented on test/e2e/prometheusagent_test.go:
```diff
@@ -579,6 +580,57 @@ func testPrometheusAgentDaemonSetSelectPodMonitor(t *testing.T) {
+	_, err = framework.CreatePrometheusAgentAndWaitUntilReady(context.Background(), ns, p)
+	require.NoError(t, err)
+
+	// The Agent doesn't support querying. So we just ensure that we are able to reach the
+	// Prometheus Agent through our service.
```
(nit) since both tests request the same API endpoint.
```suggestion
```

simonpasquier commented on test/e2e/prometheusagent_test.go:
```diff
@@ -579,6 +580,57 @@ func testPrometheusAgentDaemonSetSelectPodMonitor(t *testing.T) {
+	require.Empty(t, targets)
+
+	// Ensure that governing service was not created.
+	governingServiceName := "prometheus-operated"
```
it might be more reliable to list all the services in the namespace and check that we have only one item (same comment for the prometheus test).
```suggestion
	governingServiceName := "prometheus-agent-operated"
```

mviswanathsai replied:
Will check the number of services instead.

mviswanathsai replied:
And also added a check to ensure that the one service we have is the one we expect. This should be pretty reliable.

simonpasquier reviewed (approved):
thanks!
